### DISCUSSION TITLE
API call 'POST' for creating Pages is not working when using GITHUB_TOKEN

### DISCUSSION THREAD

**Discussion by Dawid-Bastek** [community/community#29011] [Other Feature Feedback, Questions, & Ideas] [answered]
Hello, I am getting 500 response error when trying to create Github Pages through the API with the usage of bash curl and `secrets.GITHUB_TOKEN`.
Moreover, when I am using my private access token as a hardcoded value instead of GITHUB_TOKEN, all works 🤷‍♂️ 
In the beginning, I thought maybe there was an issue with access control for GITHUB_TOKEN, and there is some missing `write` access, but when I changed [it manually](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions), it did not help 😢 

curl data:
>          curl \
>           -H "Accept: application/vnd.github+json" \
>           -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
>            https://api.github.com/repos/${{ github.repository }}/pages \
>           -d '{"source":{"branch":"gh-pages","path":"/"}}'

I used those docs as a reference: https://docs.github.com/en/rest/pages

The weird thing is that the response is blank, and when I used `-vv` command, there was no message for 500.

Logs from Github:
![image](https://user-images.githubusercontent.com/103562524/183897479-8fde59cb-f616-4eba-953b-3260b19888e9.png)

This is what Github Pages looks like after the call (you can spot the missing link to Github Pages):
<img width="865" alt="image" src="https://user-images.githubusercontent.com/103562524/185389803-fc03e42b-cf8e-4f72-b770-08dfbaac6ba6.png">

There was some post on the forum with a similar issue, but in my case, I have a problem with GITHUB_TOKEN: https://github.com/orgs/community/discussions/24343

Thanks in advance for all your help! 🙏

**Comment by Dawid-Bastek** [comment]
I saw your change on Github Pages doc with API method types, but it is not working in my case.
Now I am trying to call API via:

>            curl \
>            -X POST \
>            -H "Accept: application/vnd.github+json" \
>            -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
>            https://api.github.com/repos/${{ github.repository }}/pages \
>            -d '{"source":{"branch":"gh-pages","path":"/"}}'

And still no luck 😭 
<img width="1244" alt="image" src="https://user-images.githubusercontent.com/103562524/185387308-377ea959-1f7f-4a35-b15e-b5d5fcdbf638.png">

Any help from Github team, please?
@zackfern @davidstaheli @courtneycl @jeffbeehler

**Reply by yoannchaudet** [reply]
👋 To quote the documentation:

> Users must have write permissions. GitHub Apps must have the `pages:write` permission to use this endpoint.

By default the `GITHUB_TOKEN` does not have this permission set. You can fix that by adding a `permissions` section in your workflow with this definition: https://github.com/actions/starter-workflows/blob/0acc995eecf6f1d18190fc72ac2707b6a23d69dd/pages/jekyll.yml#L15.

**Comment by Dawid-Bastek** [comment]
Thanks @yoannchaudet !
I will check this, but I thought it is the same function and setting this up is enough:
![image](https://user-images.githubusercontent.com/103562524/186602407-a3324921-33b3-4752-890e-06cbda50448e.png)

**Comment by Dawid-Bastek** [comment]
@yoannchaudet, no luck again 😕 
<img width="1284" alt="image" src="https://user-images.githubusercontent.com/103562524/186658786-401fd34e-42e7-46bf-ae49-162cca003b39.png">


I added permissions as you mentioned:
<img width="726" alt="image" src="https://user-images.githubusercontent.com/103562524/186658609-b7dc06ba-7c30-4000-9e0c-3fb327cb715f.png">

Any more tips? 🤔

**Reply by yoannchaudet** [reply]
Oh I missed that it was a 500 error the first time. Let me get back to you when at a computer. That should not happen.

**Reply by yoannchaudet** [reply]
I opened a bug on our side so it can be fixed. The API does return a 5xx and it should not but your Pages site is enabled properly. Pages build happens with Actions nowadays and the `GITHUB_TOKEN` does not trigger subsequent workflows (for security) so your site would not be built without a consecutive commit to your `gh-pages` branch.

Actually I suggest if possible you not only enable Pages with Actions but build and deploy your site with Actions directly (it is more straightforward). You can learn about that here: https://github.blog/changelog/2022-07-27-github-pages-custom-github-actions-workflows-beta/. We have [starter workflows](https://github.com/actions/starter-workflows/tree/main/pages) doing all the work (including enabling Pages as needed).

We will update this post when we publish a fix for the 5xx status code.

**Reply by Dawid-Bastek** [reply]
Thanks for the info. I will be waiting for an update 🙂

**Reply by Dawid-Bastek** [reply]
@yoannchaudet, any updates?

**Reply by yoannchaudet** [reply]
We did make a fix for that but a security bug bounty came at roughly the same time and took precedence. While the 5xx should be gone you will probably get a 401 error trying to call the create/delete Pages API without a personal access token. The `pages:write` permission is not sufficient enough for this operation.

**Comment by JamesMGreene** [comment, marked as answer]
@Dawid-Bastek The fix @yoannchaudet [mentioned](https://github.com/orgs/community/discussions/29011#discussioncomment-3696201) has been shipped. 🚀 

Specifically, the behavior of attempting to create or delete a Pages site via the API with an Actions-generated `GITHUB_TOKEN` will now result in a `403` (Forbidden) response status with a response body similar to the following:

```json
{
  "message": "Resource not accessible by integration",
  "documentation_url": "https://docs.github.com/rest/pages#create-a-github-pages-site"
}
```

We understand that probably wasn't the fix you were hoping for, but, unfortunately, it was the one deemed necessary based on other recently raised security concerns. 😕
